### PR TITLE
[Bug Fix]Acquire lease on SPC for add event pool creation

### PR DESCRIPTION
**Pull request by sonasingh46**
This PR will do following:

    1. Add code for lease logic.
    2. Before creating a storagepool in
       case of add event of spc, lease will
       be acquired so that no two replica
       process the same request.

**Newly Added Files**
1.cmd/maya-apiserver/spc-watcher/lease.go
2.cmd/maya-apiserver/spc-watcher/lease_test.go

**Modified File**
1.cmd/maya-apiserver/spc-watcher/handler.go

**Env Variables Used**
1. "OPENEBS_MAYA_POD_NAME"
2.  "OPENEBS_NAMESPACE"
The PR is dependent on following PR (as it uses env variables exposed by downward api):
https://github.com/openebs/openebs/pull/1870

Signed-off-by: sonasingh46 <sonasingh46@gmail.com>

### CODE REVIEW

File: cmd/maya-apiserver/spc-watcher/lease_test.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	openebsFakeClientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset/fake"
+
+	"testing"
+	"github.com/golang/glog"
+)
+
+// SpcCreator will create face spc objects
```
Comment (sonasingh46) [inline cmd/maya-apiserver/spc-watcher/lease_test.go]: Will fix the typo

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	// SpcLeaseValue is the value of the SpcLeaseKey
+	SpcLeaseValue = "acquired"
+)
+
+// Leases is an interface which assists in getting and releasing lease over an spc object
+type Leases interface {
+	// GetLease will try to get a lease on spc, in case of failure it will return error
+	GetLease() (string, error)
+	// UpdateLease will update the lease value of the spc
+	UpdateLease() (string, error)
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/spc-watcher/lease.go]: I guess, interface methods & struct methods do not match.
However, will it be possible/better to have different interface or higher order function for each CRUD operation.

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2017 The OpenEBS Authors
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease.go]: 2018

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	"strings"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
+
+const (
+	// SpcLeaseKey is the key that will we used to acquire lease on spc object.
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease.go]: will `be`

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		newSpcObject.Annotations = mapCastName
+
+	} else {
+		newSpcObject.Annotations[sl.leaseKey] = leaseValue
+	}
+	spcObject, err := sl.oecs.OpenebsV1alpha1().StoragePoolClaims().Update(sl.spcObject)
+	if err != nil {
+		return nil, "", err
+	}
+	return spcObject, spcObject.Annotations[sl.leaseKey], nil
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease.go]: spcObject might be enough, specific annotation can be extracted from it.

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
+
+const (
+	// SpcLeaseKey is the key that will we used to acquire lease on spc object.
+	// It will be present in spc annotations
+	// If key has an empty value, that means no one has acquired a lease on spc object.
+	SpcLeaseKey = "openebs.io/spc-lease"
+	// SpcLeaseValue is the value of the SpcLeaseKey
+	SpcLeaseValue = "acquired"
```
[the remark this answers is not in the document]
Comment (princerachit) [reply cmd/maya-apiserver/spc-watcher/lease.go]: I agree with @AmitKumarDas There should be a unique value to distinguish who has acquired lease.

File: cmd/maya-apiserver/spc-watcher/lease_test.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2017 The OpenEBS Authors
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease_test.go]: 2018

File: cmd/maya-apiserver/spc-watcher/lease_test.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+	// Key of map describes test case behaviour.
+	// Value of map is the test object.
+	tests := map[string]struct {
+		// fakestoragepoolclaim holds the fake storagepoolcalim object in test cases.
+		fakestoragepoolclaim *apis.StoragePoolClaim
+		// expectedResult holds the expected result for the test case under run.
+		expectedResult string
+	}{
+		// TestCase#1
+		"SPC#1": {
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease_test.go]: Little description like. `lease already acquired`, `lease not acquired by any pod` etc would be good

File: cmd/maya-apiserver/spc-watcher/lease_test.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+			fakestoragepoolclaim: focs.SpcCreator("pool1", true, "acquired"),
+			expectedResult:       "",
+		},
+
+		// TestCase#2
+		"SPC#2": {
+			fakestoragepoolclaim: focs.SpcCreator("pool2", false, ""),
+			expectedResult:       "acquired",
+		},
+		// TestCase#1
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease_test.go]: TestCase#3

File: cmd/maya-apiserver/spc-watcher/lease_test.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+			fakestoragepoolclaim: focs.SpcCreator("pool1", true, "acquired"),
+			expectedResult:       "",
+		},
+
+		// TestCase#2
+		"SPC#2": {
+			fakestoragepoolclaim: focs.SpcCreator("pool2", false, ""),
+			expectedResult:       "",
+		},
+		// TestCase#1
```
Comment (princerachit) [inline cmd/maya-apiserver/spc-watcher/lease_test.go]: case 3

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 5 above)
Comment (avishnu) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Interface method return values are not matching the declared method's return values.

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+		return newLeaseValue, nil
+	}
+	// If leaseValue is not empty, lease cannot be acquired.
+	return "", fmt.Errorf("lease on spc already acquried")
+}
+
+func (sl *spcLease) UpdateLease(leaseValue string) (*apis.StoragePoolClaim, string, error) {
+	newSpcObject := sl.spcObject
+	if newSpcObject.Annotations == nil {
+		// make a map that should contain the castemplate name
```
Comment (avishnu) [inline cmd/maya-apiserver/spc-watcher/lease.go]: Where does cas template come into picture here?

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (avishnu) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Can the SpcLeaseValue be the pod name or node name instead of the string "Acquired" ?

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		return nil, "", err
+	}
+	return spcObject, spcObject.Annotations[sl.leaseKey], nil
+}
+// Can be used for reconcile loop use cases
+// TODO Remove using patch instead of update
+func (sl *spcLease) RemoveLease() (*apis.StoragePoolClaim) {
+	spcObject, _, err := sl.UpdateLease("")
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Lease could not be removed:%v", err))
```
Comment (avishnu) [inline cmd/maya-apiserver/spc-watcher/lease.go]: Lease should be lease ?

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: I used release lease using defer statement, buth then I scrapped the idea for now.
There is TODO in removeLease.
Will do after that.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 12 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: We are logging it,  not returning

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 10 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: Will fix it

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 9 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: Will fix it

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 8 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: That will give a value. I will add it.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 11 above)
Comment (avishnu) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Check the variable name as well.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 5 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: @princerachit Had both opinions, just went by the one I put.
Any way annotations need to be extracted. The idea was to make the user of function avoid any calculations for getting the annotation on which the current operation is being done.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 11 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Yes vishnu

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 2 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: I will fix it.

Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: About higher order functions . We can discuss.
If it gives us benefit for our use case I will implement it.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 5 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: @avishnu I am fixing the mismatch of method returns.

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	if strings.TrimSpace(leaseValue) == "" {
+		_, newLeaseValue, err := sl.UpdateLease(SpcLeaseValue)
+		if err != nil {
+			return "", err
+		}
+		return newLeaseValue, nil
+	}
+	// If leaseValue is not empty, lease cannot be acquired.
+	return "", fmt.Errorf("lease on spc already acquried")
+}
```
[the remark this answers is not in the document]
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Will fix it.

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+*/
+package spc
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	"strings"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
```
[the remark this answers is not in the document]
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Sure... BTW I think I have done it.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 11 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Fixed

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 10 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 9 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 7 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 8 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 5 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 4 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 3 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 2 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Fixed. But about 3 separate interfaces,we will see it in upcoming PR after more discussions.

File: cmd/maya-apiserver/spc-watcher/lease_test.go (hunk 1 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease_test.go]: fixed

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+const (
+	// SpcLeaseKey is the key that will be used to acquire lease on spc object.
+	// It will be present in spc annotations
+	// If key has an empty value, that means no one has acquired a lease on spc object.
+	SpcLeaseKey = "openebs.io/spc-lease"
```
Comment (sonasingh46) [inline cmd/maya-apiserver/spc-watcher/lease.go]: Make SpcLeaseKey to ```openebs.io/spc-lease```

File: cmd/maya-apiserver/spc-watcher/handler.go [hunk 16]
```diff
@@ -70,7 +70,13 @@ func (c *Controller) spcEventHandler(operation string, spcGot *apis.StoragePoolC
 		// CreateStoragePool function will create the storage pool
 		// It is a create event so resync should be false and pendingPoolcount is passed 0
 		// pendingPoolcount is not used when resync is false.
-		err := CreateStoragePool(spcGot, false, 0)
+		var newSpcLease Leases
+		newSpcLease = &spcLease{spcGot, SpcLeaseKey, c.clientset}
+		_, err := newSpcLease.GetLease()
+		if err != nil {
+			return addEvent, fmt.Errorf("could not acquire lease on spc object:%v", err)
+		}
```
Comment (avishnu) [inline cmd/maya-apiserver/spc-watcher/handler.go]: @sonasingh46 , can you once check the function processNextWorkItem() from runner.go and its return type. I'm trying to figure out where the error is finally logged and it seems it has to be logged inside processNextWorkItem().

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Lease could not be removed:%v", err))
+	}
+	return spcObject
+}
+
+func (sl *spcLease) getPodName() string {
+	podName := env.Get(PodNameEnvKey)
+	podNameSpace := env.Get(PodNameSpaceEnvKey)
+	return podNameSpace + "/" + podName
```
Comment (avishnu) [inline cmd/maya-apiserver/spc-watcher/lease.go]: Good !!

File: cmd/maya-apiserver/spc-watcher/handler.go (hunk 16 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/handler.go]: sure

Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/handler.go]:
```
// Run the syncHandler, passing it the namespace/name string of the
		//SPC resource to be synced.
		if err := c.syncHandler(q.Key, q.Operation, q.Object); err != nil {
			return fmt.Errorf("Error syncing '%s': %s", q.Key, err.Error())
		}
		glog.V(1).Infof("Successfully synced '%s'", q.Key)
		return nil
	}(obj)

	if err != nil {
		runtime.HandleError(err)
		return true
	}
```
In above snippet , error is not logged, it is handled runtime in if block.
Logging will cause a flood. But if you say, i can put a level log there inside the if block.
@avishnu

Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/handler.go]: fixed.... by logging the error

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+	leaseValue := sl.spcObject.Annotations[sl.leaseKey]
+	// If leaseValue is empty acquire lease.
+	if strings.TrimSpace(leaseValue) == "" {
+		spcObject, err := sl.UpdateLease(sl.getPodName())
+		if err != nil {
+			return "", err
+		}
+		return spcObject.Annotations[sl.leaseKey], nil
+	}
+	// If leaseValue is not empty, lease cannot be acquired.
```
Comment (kmova) [inline cmd/maya-apiserver/spc-watcher/lease.go]: What happens if the lease is taken by m-apiserver and it crashes before completing the SPC creation process? Sounds like we will need a lease-expiry mechanism, probably based on the time/presence of the pod that took the lease.

File: cmd/maya-apiserver/spc-watcher/lease.go [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+const (
+	// SpcLeaseKey is the key that will be used to acquire lease on spc object.
+	// It will be present in spc annotations
+	// If key has an empty value, that means no one has acquired a lease on spc object.
+	SpcLeaseKey = "openebs.io/spc-create-lease"
+	// PodNameEnvKey is the key to fetch name of the pod,which will be combined with PodNameSpaceEnvKey
+	// to be used as a value of SpcLeaseKey.
+	// e.g. "openebs.io/spc-lease":"openebs/maya-apiserver-6b4695c9f8-nbwl9"
+	PodNameEnvKey = "OPENEBS_MAYA_POD_NAME"
```
[the remark this answers is not in the document]
Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Do you mean to have a separate file having all the env keys related to maya apiserver and then use them in entire project where ever required?

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 18 above)
Comment (kmova) [reply cmd/maya-apiserver/spc-watcher/lease.go]: When does a resync event get triggered? How often does it get triggered?

Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: A resync event will be triggered after a fixed interval of time( for us it is 20 or 30 seconds, not remembering) and we can customize this.
So resync event is basically a update hook that watcher keeps on firing by delevering the old and new spc objects and you can compare the objects etc to  hook in your handlers to deal with the event.

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (kmova) [reply cmd/maya-apiserver/spc-watcher/lease.go]: This one. https://github.com/openebs/maya/blob/master/pkg/env/v1alpha1/env.go

File: cmd/maya-apiserver/spc-watcher/lease.go (hunk 18 above)
Comment (kmova) [reply cmd/maya-apiserver/spc-watcher/lease.go]: cool. So with the current code, what happens in the below scenario:
- SPC is created by kubectl apply -- for  creating pools with disks. 
- m-apiserver gets the lease, but crashes immediately after due to either node failure or another issue. 
- m-apiserver gets re-launched to another node with a new pod name. 
- SPC create event is received by the (new) m-apiserver
--
Need clarification on the following:
(a) It appears that (new) m-apiserver will fail to get lease because the older one was never removed.  To recover from this state, the SPC need to be deleted and re-added. 
(b) When the (new) m-apiserver gets the resync, does it go ahead and create it?

Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: **(a) It appears that (new) m-apiserver will fail to get lease because the older one was never removed.  To recover from ---this state, the SPC need to be deleted and re-added.** 

--->Deletion of spc should not be done from maya side IMO. It is a user created resource.
Just how happens PVC -> PV ( Does PV creation failure demands a PVC recreate ? ) 
Our case has bit lengthy chain , SPC->CSP->DEPLOY->SP

**(b) When the (new) m-apiserver gets the resync, does it go ahead and create it?**

--->When the new m-apiserver gets the resync, it does not check for the lease annotation but directly go to resync logic where it figures out current state of world which in our case for now is the pool count. The current pool count for a spc is the count of cstorpool cr that is currently existing.
Then if the desired state of world which is maxPool( specified in spc) is not matching to current state, it will go and call the CreatePool logic ( which has the nodeDiskAlloter mechanism for figuring out the resources to provision the spare amount of pool) and then will try to converge to desired state by creating pools.
So in short resync will always go and try to converge irrespective of spc lease.

But this works out when we have a single instance of maya-pod, for multiple instance there can be point of failure in resync in following case. Let us consider an example wehre maya pod is replicated to a count 2.
1. DesiredPoolCount=6
2. CurrentPoolCount=3
resyncRecieved ---- > mayaPod will try to create 3 spare pool , now if 2 maya pods reads the csp count at same time, all will go and try to converge and we will end having 6 pool resources. This is undesirable !!

But still trying to think when can a resync be faulty even in 1 instance of may pod!

Comment (sonasingh46) [reply cmd/maya-apiserver/spc-watcher/lease.go]: Short Story : 
1. Manual Provisioned SPC are not reconciled.
2. SPC lease is significant only in case of create event.
3. Resync will always happen for auto provisioned SPCs by dishonouring the SPC lease.( Problem if maya pod instance>1)
Certainly needs improvement.

Comment (kmova) [reply cmd/maya-apiserver/spc-watcher/lease.go]: thank you!  :-)
